In [1]:
!pip install transformers  #installs the hugging face transformers liabrary and provides pre-trained models like BERT, GPT, T5, mBART etc.

!pip install transformers sentencepiece  #"sentencepiece" is a tokenizer for handling sub-word tokenization

!pip install transformers datasets  #"datasets" is a library from Hugging Face for downloading and handling large datasets easily.
#Supports datasets for text, images, and audio.

!pip install datasets  #Installs only the datasets library (without transformers). Allows loading datasets from Hugging Face.

!pip install sacrebleu  #A library for evaluating machine translation models.

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 485.4/485.4 kB 6.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 3.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 6.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.8/194.8 kB 4.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.8/51.8 kB 1.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.1/104.1 kB 4.2 MB/s eta 0:00:00


In [2]:
from transformers import pipeline, AutoTokenizer, AutoModelForSeq2SeqLM

#"pipeline" is a high-level API for performing NLP tasks with a single line of code.
#You don’t need to manually load models or tokenizers—it does everything for you.
#Supports tasks like translation, text summarization, sentiment analysis, etc..

#"AutoTokenizer" loads the correct tokenizer for any given transformer model.
#A tokenizer converts text into numerical tokens before feeding it to a model.
#Automatically selects the correct tokenization technique based on the model.

#"AutoModelForSeq2SeqLM" loads a pre-trained sequence-to-sequence model (used for translation, summarization, etc.).
#Unlike pipeline(), this gives full control over model usage.
#You need to manually tokenize input, pass it to the model, and decode the output.


import pickle

import pandas as pd

import csv

from datasets import load_dataset

import matplotlib.pyplot as plt

import seaborn as sns

### **Loading the Flores-200 dataset for English Langauge:**

In [3]:
#Specifying the dataset name and configuration
dataset_name = "Muennighoff/flores200"
config_name = "eng_Latn"

#Loading the dataset
dataset = load_dataset(dataset_name, config_name)

#Choosing the appropriate slit key
split_key = "dev"  #We can set this to "devtest" as well if we prefer that split

selected_split = dataset[split_key]

eng_data = pd.DataFrame(selected_split)

print(eng_data.head())

#Saving the dataframe to a csv file
csv_filename = "flores200_data_english.csv"
eng_data.to_csv(csv_filename, index=False)
print(f"Data has been saved to the {csv_filename} file")

#Reading the csv file as a pandas dataframe
eng_data = pd.read_csv(csv_filename)
print(eng_data.head())
print(eng_data.columns)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/9.01k [00:00<?, ?B/s]

flores200.py:   0%|          | 0.00/10.5k [00:00<?, ?B/s]

The repository for Muennighoff/flores200 contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/Muennighoff/flores200.
You can avoid this prompt in future by passing the argument `trust_remote_code=True`.

Do you wish to run the custom code? [y/N] y


Generating dev split: 0 examples [00:00, ? examples/s]

Generating devtest split: 0 examples [00:00, ? examples/s]

   id                                                URL    domain     topic  \
0   1  https://en.wikinews.org/wiki/Scientists_say_ne...  wikinews    health   
1   2  https://en.wikinews.org/wiki/Scientists_say_ne...  wikinews    health   
2   3  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   
3   4  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   
4   5  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   

   has_image  has_hyperlink                                           sentence  
0          0              0  On Monday, scientists from the Stanford Univer...  
1          0              0  Lead researchers say this may bring early dete...  
2          0              0  The JAS 39C Gripen crashed onto a runway at ar...  
3          0              0  The pilot was identified as Squadron Leader Di...  
4          0              0  Local media reports an airport fire vehicle ro...  
Data has been saved to the flores

In [4]:
columns_of_interest = ["id", "URL", "domain", "topic", "has_image", "has_hyperlink"]

unique_value_counts = eng_data[columns_of_interest].nunique()  #.nunique() function is used to count the number of unique values in a DataFrame or Series.

print(unique_value_counts)

id               997
URL              281
domain             3
topic            172
has_image          1
has_hyperlink      1
dtype: int64


In [5]:
eng_data.shape

(997, 7)

### **Loading the Flores-200 dataset for the Hindi Langauge:**

In [6]:
#Specifying the dataset name and configuration
dataset_name = "Muennighoff/flores200"
config_name = "hin_Deva"

#Loading the dataset
dataset = load_dataset(dataset_name, config_name)

#Choosing the appropriate slit key
split_key = "dev"  #We can set this to "devtest" as well if we prefer that split

selected_split = dataset[split_key]

hindi_data = pd.DataFrame(selected_split)

print(hindi_data.head())

#Saving the dataframe to a csv file
csv_filename = "flores200_data_hindi.csv"
hindi_data.to_csv(csv_filename, index=False)
print(f"Data has been saved to the {csv_filename} file")

#Reading the csv file as a pandas dataframe
hindi_data = pd.read_csv(csv_filename)
print(hindi_data.head())
print(hindi_data.columns)

Generating dev split: 0 examples [00:00, ? examples/s]

Generating devtest split: 0 examples [00:00, ? examples/s]

   id                                                URL    domain     topic  \
0   1  https://en.wikinews.org/wiki/Scientists_say_ne...  wikinews    health   
1   2  https://en.wikinews.org/wiki/Scientists_say_ne...  wikinews    health   
2   3  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   
3   4  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   
4   5  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   

   has_image  has_hyperlink                                           sentence  
0          0              0  सोमवार को, स्टैनफ़ोर्ड यूनिवर्सिटी स्कूल ऑफ़ म...  
1          0              0  शोधकर्ताओं ने कहा है कि यह अल्प आय वाले देशों ...  
2          0              0  स्थानीय समय (0230 UTC) के मुताबिक करीब 9:30 बज...  
3          0              0  पायलट की पहचान स्क्वाड्रन लीडर दिलोकृत पटावी क...  
4          0              0  स्थानीय मीडिया ने बताया है कि कार्रवाई करने के...  
Data has been saved to the flores

In [7]:
columns_of_interest = ["id", "URL", "domain", "topic", "has_image", "has_hyperlink"]

unique_value_counts = hindi_data[columns_of_interest].nunique()

print(unique_value_counts)

id               997
URL              281
domain             3
topic            172
has_image          1
has_hyperlink      1
dtype: int64


### **Merging the two dataframes:**

In [8]:
merged_data = pd.merge(eng_data, hindi_data, on=["id", "domain", "topic"])

merged_data.columns

Index(['id', 'URL_x', 'domain', 'topic', 'has_image_x', 'has_hyperlink_x',
       'sentence_x', 'URL_y', 'has_image_y', 'has_hyperlink_y', 'sentence_y'],
      dtype='object')

In [9]:
#Renaming the relevant columns
merged_data.rename(columns = {"sentence_x" : "English_Sentence", "sentence_y" : "Hindi_Sentence"}, inplace=True)

merged_data.columns

Index(['id', 'URL_x', 'domain', 'topic', 'has_image_x', 'has_hyperlink_x',
       'English_Sentence', 'URL_y', 'has_image_y', 'has_hyperlink_y',
       'Hindi_Sentence'],
      dtype='object')

In [10]:
#Saving the new dataframe to a csv file
merged_data.to_csv("merged_flores200_data.csv", index=False)

In [11]:
#Re-loading the data again fro, the saved csv file
merged_data = pd.read_csv("merged_flores200_data.csv")

merged_data.head()

,id,URL_x,domain,topic,has_image_x,has_hyperlink_x,English_Sentence,URL_y,has_image_y,has_hyperlink_y,Hindi_Sentence
0,1,https://en.wikinews.org/wiki/Scientists_say_ne...,wikinews,health,0,0,"On Monday, scientists from the Stanford Univer...",https://en.wikinews.org/wiki/Scientists_say_ne...,0,0,"सोमवार को, स्टैनफ़ोर्ड यूनिवर्सिटी स्कूल ऑफ़ म..."
1,2,https://en.wikinews.org/wiki/Scientists_say_ne...,wikinews,health,0,0,Lead researchers say this may bring early dete...,https://en.wikinews.org/wiki/Scientists_say_ne...,0,0,शोधकर्ताओं ने कहा है कि यह अल्प आय वाले देशों ...
2,3,https://en.wikinews.org/wiki/Fighter_jet_crash...,wikinews,accident,0,0,The JAS 39C Gripen crashed onto a runway at ar...,https://en.wikinews.org/wiki/Fighter_jet_crash...,0,0,स्थानीय समय (0230 UTC) के मुताबिक करीब 9:30 बज...
3,4,https://en.wikinews.org/wiki/Fighter_jet_crash...,wikinews,accident,0,0,The pilot was identified as Squadron Leader Di...,https://en.wikinews.org/wiki/Fighter_jet_crash...,0,0,पायलट की पहचान स्क्वाड्रन लीडर दिलोकृत पटावी क...
4,5,https://en.wikinews.org/wiki/Fighter_jet_crash...,wikinews,accident,0,0,Local media reports an airport fire vehicle ro...,https://en.wikinews.org/wiki/Fighter_jet_crash...,0,0,स्थानीय मीडिया ने बताया है कि कार्रवाई करने के...


In [12]:
merged_data.columns

Index(['id', 'URL_x', 'domain', 'topic', 'has_image_x', 'has_hyperlink_x',
       'English_Sentence', 'URL_y', 'has_image_y', 'has_hyperlink_y',
       'Hindi_Sentence'],
      dtype='object')

In [13]:
#Dropping unnecessary columns
merged_data.drop(columns = ["URL_x", "has_image_x", "has_hyperlink_x", "URL_y", "has_image_y", "has_hyperlink_y"], inplace=True)

In [14]:
merged_data.columns

Index(['id', 'domain', 'topic', 'English_Sentence', 'Hindi_Sentence'], dtype='object')

### **Downloading the "Helsinki-NLP/opus-mt-en-hi" machine translation model:**

In [23]:
translator_opus = pipeline("translation_en_to_hi", model="Helsinki-NLP/opus-mt-en-hi", device=-1)
#"translation_en_to_hi" is a task name to be handled by the model
#"device=-1" means we are allowing it to use the CPU and if "device=0", it means it can use GPU for the processing.

english_sentence = input("Enter an english sentence : ")

hindi_translator_opus = translator_opus(english_sentence, max_length=50)[0]['translation_text']

print("Translation (opus):", hindi_translator_opus)

Device set to use cpu


Enter an english sentence : Hello. How are you? I am fine. Where are you now?
Translation (opus): हैलो, तुम कैसे हो? मैं ठीक हूँ. अब तुम कहाँ हो?


### **Downloading the "facebook/m2m100_418M" machine translation model:**

In [29]:
from transformers import pipeline, AutoTokenizer, AutoModelForSeq2SeqLM

translator_m2m = pipeline("translation_en_to_hi", model="facebook/m2m100_418M", device=-1)

english_sent = input("Enter the english sentence here : ")

hindi_translator_m2m = translator_m2m(english_sent, max_length=50)[0]['translation_text']

print(f"Translation m2m : {hindi_translator_m2m}")

Device set to use cpu


Enter the english sentence here : Hey how are you? Where do you stay these days?
Translation m2m : आप कहां रहते हैं इन दिनों?


### **Downloading the "facebook/mbart-large-50-many-to-many-m-model" machine translation model:**

In [32]:
from transformers import pipeline, AutoTokenizer, AutoModelForSeq2SeqLM

translator_mbart = pipeline("translation", model="facebook/mbart-large-50-many-to-many-mmt", device=-1, src_lang="en_X",
                            tgt_lang="hi_IN")

english_sentence1 = input("Enter the english sentence here : ")

hindi_mbart = translator_mbart(english_sentence1, max_length=50)[0]['translation_text']

print(f"Translation (mbart): {hindi_mbart}")

config.json:   0%|          | 0.00/1.43k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.44G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/261 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/529 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/649 [00:00<?, ?B/s]

Device set to use cpu


Enter the english sentence here : I am going to Mumbai tomorrow. Would you like to come with me?
Translation (mbart): मैं कल मुंबई जा रहा हूँ। क्या आप मेरे साथ आना चाहेंगे?


In [35]:
merged_df_subset = merged_data.head()

print(merged_df_subset)

   id    domain     topic                                   English_Sentence  \
0   1  wikinews    health  On Monday, scientists from the Stanford Univer...   
1   2  wikinews    health  Lead researchers say this may bring early dete...   
2   3  wikinews  accident  The JAS 39C Gripen crashed onto a runway at ar...   
3   4  wikinews  accident  The pilot was identified as Squadron Leader Di...   
4   5  wikinews  accident  Local media reports an airport fire vehicle ro...   

                                      Hindi_Sentence  
0  सोमवार को, स्टैनफ़ोर्ड यूनिवर्सिटी स्कूल ऑफ़ म...  
1  शोधकर्ताओं ने कहा है कि यह अल्प आय वाले देशों ...  
2  स्थानीय समय (0230 UTC) के मुताबिक करीब 9:30 बज...  
3  पायलट की पहचान स्क्वाड्रन लीडर दिलोकृत पटावी क...  
4  स्थानीय मीडिया ने बताया है कि कार्रवाई करने के...  


In [36]:
#Translating English sentences to hindi
merged_df_subset["Hindi_opus"] = merged_df_subset['English_Sentence'].apply(translator_opus)

merged_df_subset["Hindi_m2m"] = merged_df_subset['English_Sentence'].apply(translator_m2m)

merged_df_subset["Hindi_mbart"] = merged_df_subset['English_Sentence'].apply(translator_mbart)

<ipython-input-36-893d60c2600e>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df_subset["Hindi_opus"] = merged_df_subset['English_Sentence'].apply(translator_opus)
<ipython-input-36-893d60c2600e>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df_subset["Hindi_m2m"] = merged_df_subset['English_Sentence'].apply(translator_m2m)
<ipython-input-36-893d60c2600e>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer]

In [37]:
merged_df_subset

,id,domain,topic,English_Sentence,Hindi_Sentence,Hindi_opus,Hindi_m2m,Hindi_mbart
0,1,wikinews,health,"On Monday, scientists from the Stanford Univer...","सोमवार को, स्टैनफ़ोर्ड यूनिवर्सिटी स्कूल ऑफ़ म...",[{'translation_text': 'यह एक छोटी - सी किस्म क...,"[{'translation_text': 'सोमवार को, स्टैनफोर्ड य...","[{'translation_text': 'Monday, scientists from..."
1,2,wikinews,health,Lead researchers say this may bring early dete...,शोधकर्ताओं ने कहा है कि यह अल्प आय वाले देशों ...,[{'translation_text': 'सबसे आगे के खोजकर्ताओं ...,[{'translation_text': 'प्रमुख शोधकर्ताओं का कह...,[{'translation_text': 'प्रमुख अनुसंधानकर्ताओं ...
2,3,wikinews,accident,The JAS 39C Gripen crashed onto a runway at ar...,स्थानीय समय (0230 UTC) के मुताबिक करीब 9:30 बज...,[{'translation_text': 'JS 39C ने 930 के आस - प...,[{'translation_text': 'JAS 39C Gripen 9:30 बजे...,[{'translation_text': ' JAS 39C Gripen ने लगभग...
3,4,wikinews,accident,The pilot was identified as Squadron Leader Di...,पायलट की पहचान स्क्वाड्रन लीडर दिलोकृत पटावी क...,[{'translation_text': 'इस विमान के पायलट को स्...,[{'translation_text': 'पायलट को स्क्वाड्रन लीड...,[{'translation_text': 'इस पायलट को स्क्वाड्रन ...
4,5,wikinews,accident,Local media reports an airport fire vehicle ro...,स्थानीय मीडिया ने बताया है कि कार्रवाई करने के...,[{'translation_text': 'स्थानीय मीडिया रिपोर्ट ...,[{'translation_text': 'स्थानीय मीडिया रिपोर्ट ...,[{'translation_text': 'स्थानीय मीडिया ने एक हव...


In [39]:
first_row = merged_df_subset.iloc[0]

for column, value in first_row.items():
  print(f"{column} : {value}")

id : 1
domain : wikinews
topic : health
English_Sentence : On Monday, scientists from the Stanford University School of Medicine announced the invention of a new diagnostic tool that can sort cells by type: a tiny printable chip that can be manufactured using standard inkjet printers for possibly about one U.S. cent each.
Hindi_Sentence : सोमवार को, स्टैनफ़ोर्ड यूनिवर्सिटी स्कूल ऑफ़ मेडिसिन के वैज्ञानिकों ने एक नए डायग्नोस्टिक उपकरण के आविष्कार की घोषणा की जो कोशिकाओं को उनके प्रकार के आधार पर छाँट सकता है: एक छोटी प्रिंट करने योग्य चिप जिसे स्टैण्डर्ड इंकजेट प्रिंटर का उपयोग करके लगभग एक अमेरिकी सेंट के लिए निर्मित किया जा सकता है.
Hindi_opus : [{'translation_text': 'यह एक छोटी - सी किस्म की छपाई मशीन है, जो अमरीका के हिसाब से अमरीका के हिसाब से अमरीका के हिसाब से एक - से - ज़्यादा लोगों के लिए एक छोटा - सा छपाई मशीन तैयार की जा सकती है ।'}]
Hindi_m2m : [{'translation_text': 'सोमवार को, स्टैनफोर्ड यूनिवर्सिटी स्कूल ऑफ मेडिसिन के वैज्ञानिकों ने एक नया निदान उपकरण का आविष्कार घोषित किया

Example 1: New Diagnostic Tool Announcement

Opus-MT Translation:
Translation appears to be nonsensical and does not convey the original meaning accurately.

m2m100 Translation:
Translation is coherent and captures the essence of the original English sentence.

mbart Translation:
The model could not translate the sentence due to which you can see that the english sentence has been printed as it is in the output.

In [41]:
first_row = merged_df_subset.iloc[2]

for column, value in first_row.items():
  print(f"{column} : {value}")

id : 3
domain : wikinews
topic : accident
English_Sentence : The JAS 39C Gripen crashed onto a runway at around 9:30 am local time (0230 UTC) and exploded, closing the airport to commercial flights.
Hindi_Sentence : स्थानीय समय (0230 UTC) के मुताबिक करीब 9:30 बजे JAS 39C Gripen रनवे पर धमाके के साथ दुर्घटनाग्रस्त हो गया, जिसकी वजह से वाणिज्यिक उड़ानों के लिए हवाई अड्डे को बंद कर दिया गया.
Hindi_opus : [{'translation_text': 'JS 39C ने 930 के आस - पास एक रनवे पर धावा बोला, जो स्थानीय समय पर होता है (0230 के दशक) और विस्फोटों से भरा हुआ था ।'}]
Hindi_m2m : [{'translation_text': 'JAS 39C Gripen 9:30 बजे स्थानीय समय (0230 यूटीसी) के आसपास एक रेलवे पर टूट गया और विस्फोट हुआ, वाणिज्यिक उड़ानों के लिए हवाई अड्डे को बंद कर दिया।'}]
Hindi_mbart : [{'translation_text': ' JAS 39C Gripen ने लगभग 9:30 बजे स्थानीय समय (01230 UTC) के 9:30 बजे एक हवाई अड्डे पर दुर्घटनाग्रस्त किया और विस्फोट कर हवाई अड्डे को वाणिज्यिक उड़ानों के लिए बंद कर दिया।'}]


Example 2: JAS 39C Gripen Crash

Opus-MT Translation:
Translation is somewhat understandable but lacks precision and fluency.

m2m100 Translation:
Translation is inaccurate and ineffectively communicates the events described in the English sentence.

mbart Translation:
Translation is accurate and provides a clear representation of the original sentence.

### **Calculating the BLEU Score for each of the above models:**

In [44]:
merged_df_subset.columns

Index(['id', 'domain', 'topic', 'English_Sentence', 'Hindi_Sentence',
       'Hindi_opus', 'Hindi_m2m', 'Hindi_mbart'],
      dtype='object')

In [51]:
from sacrebleu import corpus_bleu

original_translations = merged_df_subset['Hindi_Sentence'].tolist()

opus_translations = merged_df_subset['Hindi_opus'].tolist()
m2m_translations = merged_df_subset['Hindi_m2m'].tolist()
mbart_translations = merged_df_subset['Hindi_mbart'].tolist()

#Converting the model translations to string format
opus_translations = [str(x) for x in opus_translations]
m2m_translations = [str(x) for x in m2m_translations]
mbart_translations = [str(x) for x in mbart_translations]

#Lets calculate the BLEU Score now
bleu_opus = corpus_bleu(opus_translations, [original_translations])
bleu_m2m = corpus_bleu(m2m_translations, [original_translations])
bleu_mbart = corpus_bleu(mbart_translations, [original_translations])

print(f"BLEU Score for Opus-MT: {bleu_opus.score}")
print(f"BLEU Score for m2m100: {bleu_m2m.score}")
print(f"BLEU Score for mbart: {bleu_mbart.score}")


BLEU Score for Opus-MT: 2.256547238077094
BLEU Score for m2m100: 21.820076466821448
BLEU Score for mbart: 9.607707710221183


Opus-MT:-  2.26

A BLEU score around 2 is relatively low. Opus-MT might be struggling with the translation task, possibly due to the complexity of the sentences, domain-specific language, or other issues.

m2m100:-  21.82

A BLEU score of 21.82 is quite high, suggesting good translation quality. m2m100 is performing well on your dataset.

mbart:-  9.60

A BLEU score of 9.60 is also relatively low, indicating poor translation quality. mbart seems not to be performing well.